ext/nio4r/monitor.c

```diff
@@ -60,13 +60,18 @@ void Init_NIO_Monitor()
 static VALUE NIO_Monitor_allocate(VALUE klass)
 {
     struct NIO_Monitor *monitor = (struct NIO_Monitor *)xmalloc(sizeof(struct NIO_Monitor));
-
+    
+    if(!monitor)
+        return Qnil;
```

Comment:
I don't believe this will ever happen, as `xmalloc`'s contract is always to succeed.

`xmalloc` (via a circuitous set of wrapper functions and macros) invokes `malloc` inside of a `TRY_WITH_GC` macro wrapper that runs the GC on OOM, and if that doesn't free up sufficient memory, calls `ruby_memerror()`, which depending on circumstances will invoke `rb_memerror`. From there, it looks like one of two things can happen:

- If a specific set of conditions hold, including the GVL being held by the current thread, it will raise `NoMemoryError`. This performs a `longjmp` (or thereabouts) out of `NIO_Monitor_allocate`
- If the conditions aren't met, it will print an error message `[FATAL] failed to allocate memory` and exit the process

Nowhere in this does it ever appear to return NULL.

Reply:
True... but I don't believe we should rely on undocumented behavior of external code... at least, I didn't find such documentation fo `ruby_xmalloc`.

Your statement is actually true for the OS `malloc` on systems that implement optimistic memory allocation. i.e., [Linux's default "optimistic" memory allocation](https://www.man7.org/linux/man-pages/man3/malloc.3.html#NOTES), never returns `NULL` and only allocates the memory page once the memory address is "touched" (at which point the process would crash).

This behavior means that `if(!(ptr = malloc(sizeof(*ptr))))` is redundant on many systems. Actually, one might argue that on all non-embedded operating systems, this error test is redundant (since writing to the memory will automatically crash the process)... However, I don't think we should avoid error checks.

I'm happy to remove this from the patch, it's your call, but I don't see why we should rely on undocumented behavior.

Reply:
IMO as a C idiom, `xmalloc` is intended to be an infallible wrapper for `malloc` (or, that is to say, it implements its own strategy for handling allocation failures, even if that's crashing the process).

If it did fail, it would be equivalent to `malloc`.

Reply:
@tarcieri - I'll do whatever you think its best, just tell me "remove this" and it's done.

I understand your point of view, I just prefer to be defensive.

Reply:
It's okay to be defensive, but I'd prefer an `assert()` when the expectation/contract is for it to never be `NULL`.